File: experiment-6/runs/language=typescript_model=claude-opus-4-7_tooling=none/rep3/tests/books.test.ts
import request from 'supertest';
import type { Express } from 'express';
import { createApp } from '../src/app';
import { createDb } from '../src/db';

function makeApp(): Express {
  const db = createDb(':memory:');
  return createApp(db);
}

describe('Books API', () => {
  describe('GET /health', () => {
    it('returns 200 with ok status', async () => {
      const app = makeApp();
      const res = await request(app).get('/health');
      expect(res.status).toBe(200);
      expect(res.body).toEqual({ status: 'ok' });
    });
  });

  describe('POST /books', () => {
    it('creates a book and returns 201', async () => {
      const app = makeApp();
      const res = await request(app).post('/books').send({
        title: 'The Hobbit',
        author: 'J.R.R. Tolkien',
        year: 1937,
        isbn: '978-0345339683',
      });
      expect(res.status).toBe(201);
      expect(res.body).toMatchObject({
        id: expect.any(Number),
        title: 'The Hobbit',
        author: 'J.R.R. Tolkien',
        year: 1937,
        isbn: '978-0345339683',
      });
    });

    it('returns 400 when title is missing', async () => {
      const app = makeApp();
      const res = await request(app).post('/books').send({ author: 'Someone' });
      expect(res.status).toBe(400);
      expect(res.body.error).toMatch(/title/);
    });

    it('returns 400 when author is missing', async () => {
      const app = makeApp();
      const res = await request(app).post('/books').send({ title: 'Untitled' });
      expect(res.status).toBe(400);
      expect(res.body.error).toMatch(/author/);
    });

    it('returns 400 when year is not an integer', async () => {
      const app = makeApp();
      const res = await request(app)
        .post('/books')
        .send({ title: 'X', author: 'Y', year: 'not-a-number' });
      expect(res.status).toBe(400);
    });
  });

  describe('GET /books', () => {
    it('returns an empty list initially', async () => {
      const app = makeApp();
      const res = await request(app).get('/books');
      expect(res.status).toBe(200);
      expect(res.body).toEqual([]);
    });

    it('lists all books', async () => {
      const app = makeApp();
      await request(app).post('/books').send({ title: 'A', author: 'X' });
      await request(app).post('/books').send({ title: 'B', author: 'Y' });
      const res = await request(app).get('/books');
      expect(res.status).toBe(200);
      expect(res.body).toHaveLength(2);
    });

    it('filters by author', async () => {
      const app = makeApp();
      await request(app).post('/books').send({ title: 'A', author: 'Alice' });
      await request(app).post('/books').send({ title: 'B', author: 'Bob' });
      await request(app).post('/books').send({ title: 'C', author: 'Alice' });
      const res = await request(app).get('/books?author=Alice');
      expect(res.status).toBe(200);
      expect(res.body).toHaveLength(2);
      expect(res.body.every((b: { author: string }) => b.author === 'Alice')).toBe(true);
    });
  });

  describe('GET /books/:id', () => {
    it('returns the book when it exists', async () => {
      const app = makeApp();
      const created = await request(app)
        .post('/books')
        .send({ title: 'Dune', author: 'Frank Herbert' });
      const res = await request(app).get(`/books/${created.body.id}`);
      expect(res.status).toBe(200);
      expect(res.body.title).toBe('Dune');
    });

    it('returns 404 when the book does not exist', async () => {
      const app = makeApp();
      const res = await request(app).get('/books/9999');
      expect(res.status).toBe(404);
    });
  });

  describe('PUT /books/:id', () => {
    it('updates an existing book', async () => {
      const app = makeApp();
      const created = await request(app)
        .post('/books')
        .send({ title: 'Old', author: 'A' });
      const res = await request(app)
        .put(`/books/${created.body.id}`)
        .send({ title: 'New', author: 'A', year: 2020 });
      expect(res.status).toBe(200);
      expect(res.body.title).toBe('New');
      expect(res.body.year).toBe(2020);
    });

    it('returns 404 for missing book', async () => {
      const app = makeApp();
      const res = await request(app)
        .put('/books/9999')
        .send({ title: 'X', author: 'Y' });
      expect(res.status).toBe(404);
    });

    it('returns 400 for invalid input on update', async () => {
      const app = makeApp();
      const created = await request(app)
        .post('/books')
        .send({ title: 'T', author: 'A' });
      const res = await request(app)
        .put(`/books/${created.body.id}`)
        .send({ author: 'A' });
      expect(res.status).toBe(400);
    });
  });

  describe('DELETE /books/:id', () => {
    it('deletes an existing book', async () => {
      const app = makeApp();
      const created = await request(app)
        .post('/books')
        .send({ title: 'Gone', author: 'A' });
      const res = await request(app).delete(`/books/${created.body.id}`);
      expect(res.status).toBe(204);
      const lookup = await request(app).get(`/books/${created.body.id}`);
      expect(lookup.status).toBe(404);
    });

    it('returns 404 for missing book', async () => {
      const app = makeApp();
      const res = await request(app).delete('/books/9999');
      expect(res.status).toBe(404);
    });
  });
});

File: experiments/adrianco/experiment-8/runs/language=elixir_model=claude-opus-4-8/rep3/test/book_api/books_test.exs
defmodule BookApi.BooksTest do
  use BookApi.DataCase

  alias BookApi.Books
  alias BookApi.Book

  describe "create_book/1" do
    test "creates a book with valid attributes" do
      attrs = %{"title" => "Dune", "author" => "Frank Herbert", "year" => 1965, "isbn" => "9780441013593"}
      assert {:ok, %Book{} = book} = Books.create_book(attrs)
      assert book.title == "Dune"
      assert book.author == "Frank Herbert"
      assert book.year == 1965
    end

    test "fails when title is missing" do
      assert {:error, changeset} = Books.create_book(%{"author" => "Someone"})
      refute changeset.valid?
      assert %{title: ["can't be blank"]} = errors_on(changeset)
    end

    test "fails when author is missing" do
      assert {:error, changeset} = Books.create_book(%{"title" => "A Book"})
      refute changeset.valid?
      assert %{author: ["can't be blank"]} = errors_on(changeset)
    end
  end

  describe "list_books/1" do
    test "lists all books and filters by author" do
      {:ok, _} = Books.create_book(%{"title" => "Book A", "author" => "Alice"})
      {:ok, _} = Books.create_book(%{"title" => "Book B", "author" => "Bob"})

      assert length(Books.list_books()) == 2
      assert [%Book{author: "Alice"}] = Books.list_books(author: "Alice")
    end
  end

  describe "update_book/2 and delete_book/1" do
    test "updates a book" do
      {:ok, book} = Books.create_book(%{"title" => "Old", "author" => "Author"})
      assert {:ok, updated} = Books.update_book(book, %{"title" => "New"})
      assert updated.title == "New"
    end

    test "deletes a book" do
      {:ok, book} = Books.create_book(%{"title" => "Temp", "author" => "Author"})
      assert {:ok, _} = Books.delete_book(book)
      assert Books.get_book(book.id) == nil
    end
  end

  defp errors_on(changeset) do
    Ecto.Changeset.traverse_errors(changeset, fn {msg, opts} ->
      Enum.reduce(opts, msg, fn {key, value}, acc ->
        String.replace(acc, "%{#{key}}", to_string(value))
      end)
    end)
  end
end
